### PR TITLE
os/bluestore: make osd shard-thread do oncommits

### PR DESCRIPTION
**Pull request by majianpeng**
The benefits:
a)remove finisher threads
b)avoid racing of pg lock.
For some workload, this can decreased performance.

Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

**Comment by majianpeng** [comment]
My test envi:
one osd(P3700) as osd/rocksdb. 10 rbd image(every 10G); fio+librbd; 4k randwrite.

4k-8osdthread-1finisher: IOPS=51K
4k-8osdthread-8finisher: IOPS=51.7K
4k-8osdthread-with-pr: IOPS=49.7K
4k-16osdthread-1finisher: IOPS=50.6K
4k-16osdthread-16finisher: IOPS=54.9K
4k-16osdthread-with-patch: IOPS=58.6k.

THe idea from seastar which make callback on the same core. From the results, when workload is relative lower, has additional finisher thread has better performance. But for heavy workload, this PR show better performance(58.6/50.6=115%) and remove the additional finisher threads.

**Comment by majianpeng** [comment]
@markhpc @liewegas @ifed01 @tchaikov . please review.

**Comment by markhpc** [comment]
@majianpeng Any chance I could get you to try using the gdb wallclock profiler on a before/after test case?

You can see it here:
https://github.com/markhpc/gdbpmp

Thanks!
Mark

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -9661,224 +9662,241 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
...
-  PGRef pg = slot->pg;
-
-  // lock pg (if we have it)
-  if (pg) {
-    // note the requeue seq now...
-    uint64_t requeue_seq = slot->requeue_seq;
-    ++slot->num_running;
+  list<Context *> oncommits;
+  if (!sdata->context_queue.empty()) {
+    sdata->context_queue.swap(oncommits);
```
Comment (liewegas) [inline src/osd/OSD.cc]: Why not finish() the contexts here?

**Comment by liewegas** [comment]
This is really interesting!   A few initial thoughts:

- We can't entirely remove the finishers because sometimes the store runs outside of the OSD.
- We can come up with a cleaner way to (explicitly!) attach the context queues to the collections.
- We need to be really careful about deadlocks.  Usually they come from the throttling in queue_transaction, but I think this change won't affect anything because we release those throttles in kv_sync_thread, not the finishers.
- I think we can make the change to the op wq simpler...

Any sense of why this is slower with only 8 worker threads?

File: src/osd/OSD.cc (hunk 1 above)
Comment (majianpeng) [reply src/osd/OSD.cc]: at this point,  it already lock(sdata->shard_lock)
a)context callback may take long time
b)context will get pg lock and will cause dead lock.

**Comment by majianpeng** [comment]
@liewegas . update by your suggestion. please review.

**Comment by majianpeng** [comment]
@jdurgin .  ping. Thanks!

**Comment by majianpeng** [comment]
@jdurgin . Have you time to review this PR?  Thanks very much.

**Comment by jdurgin** [comment]
I like this idea - agree we need to be careful about deadlocks. 

The [diff --minimal](https://paste.fedoraproject.org/paste/8HbK7EEmfFc9BwDX2ZVdNA) is a little more readable - the op wq change is simple enough, the change in indentation level just makes the diff look bigger.

@majianpeng can you rebase? Then let's run this through the rados suite and see if any fallout emerges.

**Comment by majianpeng** [comment]
@liewegas @jdurgin . update  by your suggestions: a)rebase on  master ; b)change OSD::ShardedOpWQ::_process which make looks readable.    please review again. Thanks!

**Comment by majianpeng** [comment]
@liewegas . Ping. Thanks

**Comment by majianpeng** [comment]
@liewegas . Ping.

File: src/common/Finisher.h [hunk 2]
```diff
@@ -184,4 +184,40 @@ class C_OnFinisher : public Context {
...
+      cond.SignalOne();
+      mutex.Unlock();
+
+      ls.clear();
+    }
+
+    void swap(list<Context *>& ls) {
+      ls.clear();
+      std::lock_guard<std::mutex> l(q_mutex);
+      if(!q.empty()) {
```
Comment (liewegas) [inline src/common/Finisher.h]: nit: whitespace

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -9628,8 +9612,11 @@ int BlueStore::queue_transactions(
   for (auto c : on_applied_sync) {
     c->complete(0);
   }
-  for (auto c : on_applied) {
-    finishers[osr->shard]->queue(c);
+  if(on_applied.size()) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: nit: whitespace

File: src/osd/OSD.h [hunk 4]
```diff
@@ -1735,7 +1737,14 @@ class OSD : public Dispatcher,
       auto &&sdata = osd->shards[shard_index];
       assert(sdata);
       Mutex::Locker l(sdata->shard_lock);
-      return sdata->pqueue->empty();
+      return sdata->pqueue->empty() && sdata->context_queue.empty();
+    }
+
+    void handle_oncommits(list<Context*>& oncommits) {
+      if (!oncommits.empty()) {
```
Comment (liewegas) [inline src/osd/OSD.h]: Does this 'if' actually help anything?

**Comment by majianpeng** [comment]
retest this  please.

**Comment by majianpeng** [comment]
@liewegas . update by your suggestions. But  it can't pass "make check". What's mean of "Pending — Build triggered for merge commit. "?   In my dev-machine, it can pass all tests.

**Comment by liewegas** [comment]
@majianpeng The lab is down, which probably include jenkins.  SHould be fixed shortly!

**Comment by majianpeng** [comment]
retest this please

**Comment by tchaikov** [comment]
```
    -3> 2018-07-29 04:36:03.494 7fc7b12c8700  1 heartbeat_map is_healthy 'OSD::osd_op_tp thread 0x7fc794844700' had timed out after 15
    -2> 2018-07-29 04:36:03.494 7fc7b12c8700  1 heartbeat_map is_healthy 'OSD::osd_op_tp thread 0x7fc794844700' had suicide timed out after 150
    -1> 2018-07-29 04:36:03.498 7fc794844700 -1 *** Caught signal (Aborted) **
 in thread 7fc794844700 thread_name:tp_osd_tp

 ceph version 14.0.0-1698-g0f9ae53 (0f9ae53c00af3d84f5126561e07fdc7087c2af94) nautilus (dev)
 1: (()+0x11390) [0x7fc7b5dc1390]
 2: (()+0x1026d) [0x7fc7b5dc026d]
 3: (pthread_mutex_lock()+0x7d) [0x7fc7b5db9dbd]
 4: (Mutex::Lock(bool)+0x69) [0x557e79c7aa19]
 5: (OSD::_finish_splits(std::set<boost::intrusive_ptr<PG>, std::less<boost::intrusive_ptr<PG> >, std::allocator<boost::intrusive_ptr<PG> > >&)+0x48) [0x557e796d46a8]
 6: (Context::complete(int)+0x9) [0x557e797068c9]
 7: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xabc) [0x557e796fab0c]
 8: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x466) [0x557e79c8a606]
 9: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x557e79c8c180]
```

```
-10000> 2018-07-29 04:34:34.770 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c920) _trim onodes 38 / 17487 buffers 0 / 13421772
 -9999> 2018-07-29 04:34:34.818 7fc7a6868700 20 bluestore.2QCache(0x557e83f8b260) _trim onodes 234 / 17487 buffers 319488 / 13421772
 -9998> 2018-07-29 04:34:34.818 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c680) _trim onodes 38 / 17487 buffers 0 / 13421772
 -9997> 2018-07-29 04:34:34.818 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c760) _trim onodes 40 / 17487 buffers 0 / 13421772
 -9996> 2018-07-29 04:34:34.818 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c840) _trim onodes 35 / 17487 buffers 0 / 13421772
 -9995> 2018-07-29 04:34:34.818 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c920) _trim onodes 38 / 17487 buffers 0 / 13421772
... more _trim onodes ...
... and more 
    -7> 2018-07-29 04:36:03.474 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c760) _trim onodes 40 / 17487 buffers 0 / 13421772
    -6> 2018-07-29 04:36:03.474 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c840) _trim onodes 35 / 17487 buffers 0 / 13421772
    -5> 2018-07-29 04:36:03.474 7fc7a6868700 20 bluestore.2QCache(0x557e83f8c920) _trim onodes 38 / 17487 buffers 0 / 13421772
    -4> 2018-07-29 04:36:03.494 7fc7b02c6700 -1 received  signal: Hangup from /usr/bin/python /usr/bin/daemon-helper kill ceph-osd -f --cluster ceph -i 2  (PID: 54041) UID: 0
    -3> 2018-07-29 04:36:03.494 7fc7b12c8700  1 heartbeat_map is_healthy 'OSD::osd_op_tp thread 0x7fc794844700' had timed out after 15
    -2> 2018-07-29 04:36:03.494 7fc7b12c8700  1 heartbeat_map is_healthy 'OSD::osd_op_tp thread 0x7fc794844700' had suicide timed out after 150
    -1> 2018-07-29 04:36:03.498 7fc794844700 -1 *** Caught signal (Aborted) **
 in thread 7fc794844700 thread_name:tp_osd_tp
```

i am not sure it's relevant. but to be on the safe side, i am dropping this change from my test batch.

**Comment by majianpeng** [comment]
@tchaikov . I think this bug is produced by this PR. It's deadlock. But from the code, i don't find any clue.  How can do the same test on local machine?  Thanks!

**Comment by tchaikov** [comment]
@majianpeng see `/a/kchai-2018-07-28_01:50:42-rados-wip-kefu-testing-2018-07-27-2326-distro-basic-mira/2827306/remote/*/log/ceph-osd.4.log.gz`. i am not sure how to do the thrash test locally, though.

**Comment by majianpeng** [comment]
@tchaikov . I only got http://pulpito.ceph.com/kchai-2018-07-28_01:50:42-rados-wip-kefu-testing-2018-07-27-2326-distro-basic-mira/2827306/. I don't know how to get  /a/kchai-2018-07-28_01:50:42-rados-wip-kefu-testing-2018-07-27-2326-distro-basic-mira/2827306/remote/*/log/ceph-osd.4.log.gz?   Thanks!

**Comment by xiexingguo** [comment]
@majianpeng Try this **http://qa-proxy.ceph.com/teuthology/kchai-2018-07-28_01:50:42-rados-wip-kefu-testing-2018-07-27-2326-distro-basic-mira/2827306/**

**Comment by majianpeng** [comment]
@xiexingguo  Thanks very much!

**Comment by majianpeng** [comment]
@tchaikov . I can't find the reason from reading code and can't reproduce by local tests. I add a new commit which add debug message. Could you do a teuthology test  to help me find the reason?  Thanks very much.

**Comment by tchaikov** [comment]
sure.

**Comment by majianpeng** [comment]
retest this please

**Comment by majianpeng** [comment]
@tchaikov . Ping.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-08-05_01:51:41-rados-wip-kefu-testing-2018-08-05-0015-distro-basic-mira/

quite a few tests failed, i am sure some of them are not relevant . will rerun them once this batch finishes.

**Comment by majianpeng** [comment]
@liewegas .  if there are multi threads_of_shard can cause out of order for oncommit.  So i add https://github.com/ceph/ceph/pull/22739/commits/471fd772adfbe8b9fc514e36cbf8922b41538ff3.  Can you review this ? Thanks!

**Comment by liewegas** [comment]
Limiting the shards seems like the obvious workaround.  I think it changes the decision a bit about whether this overall path makes sense, though.  Right now a blocking read will block an entire PG.  I'm not sure if 1 thread per shard changes the overall performance too... we probably need to repeat some of the testing with this change.  @markhpc thoughts?

**Comment by majianpeng** [comment]
@liewegas . If using seastar,  are we also make shard_per_shard?  or the alternative way is make a thread of threads_of_shard do oncommits, etc make the  smallest thread_index(thread_index%shard_num==0) to do this(thread_index can't change after create).

**Comment by majianpeng** [comment]
@liewegas . If make threads_per_shard=1, we can increase shards to improve performance. Yes this  a big change and make user know this change.

**Comment by liewegas** [comment]
Yeah, we can probably adjust here.  One contributing problem is that reads are blocking (we don't have async reads in ObjectStore currently).  Although I'm not sure why just increasing shards by the thread factor wouldn't always be better....

**Comment by majianpeng** [comment]
@liewegas . update: because currently code already use thread_index % num_shard to choose shard. So only using thrad_index < num_shard as the smallest thread of shard. please see https://github.com/ceph/ceph/pull/22739/commits/30db4586350c776cfc6580bd6fd92100b01214c3.

**Comment by liewegas** [comment]
ah, that works!

**Comment by tchaikov** [comment]
@majianpeng needs rebase.

**Comment by majianpeng** [comment]
retest this please

**Comment by liewegas** [comment]
A few days ago I'm pretty sure this caused the failures here: http://pulpito.ceph.com/sage-2018-08-10_21:50:39-rados-wip-sage-testing-2018-08-10-1500-distro-basic-smithi/

**Comment by majianpeng** [comment]
@liewegas . I can't find any useful message (like dump-stack). And i found for all dead cases all remote data don't appear. So could you give some more suggestion for the current questions? and confirm those tests based on latest code?  Thanks very much!

**Comment by majianpeng** [comment]
@tchaikov .  From  http://pulpito.ceph.com/sage-2018-08-10_21:50:39-rados-wip-sage-testing-2018-08-10-1500-distro-basic-smithi/,  all dead cases can't contain "remote" data. So i hope you can do additional test which help me find the deadlock problem. Thanks very much.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-08-27_08:20:01-rados-wip-kefu-testing-2018-08-27-1202-distro-basic-mira/

**Comment by majianpeng** [comment]
I found the same bug in  many fail or dead case.
/build/ceph-14.0.0-2597-g5c0c839/src/mon/OSDMonitor.cc: In function 'bool OSDMonitor::do_prune(MonitorDBStore::TransactionRef)' thread 7f414c926700 time 2018-08-27 14:18:15.801030
build/ceph-14.0.0-2597-g5c0c839/src/mon/OSDMonitor.cc: 2011: FAILED assert(map_exists(last_pinned))
ceph version 14.0.0-2597-g5c0c839 (5c0c839f18e6047f062d8b9ccd1720515c62f365) nautilus (dev)
1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f4155dfb755]
2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f4155dfb8d7]
3: (OSDMonitor::do_prune(std::shared_ptr<MonitorDBStore::Transaction>)+0x1194) [0x55e43b18eb84]
4: (OSDMonitor::encode_pending(std::shared_ptr<MonitorDBStore::Transaction>)+0x77) [0x55e43b18ec17]
5: (PaxosService::propose_pending()+0x10b) [0x55e43b14b57b]
6: (OSDMonitor::tick()+0x588) [0x55e43b194848]
7: (Monitor::tick()+0x90) [0x55e43b0083d0]
8: (C_MonContext::finish(int)+0x39) [0x55e43afe3579]
9: (Context::complete(int)+0x9) [0x55e43b01c3a9]
10: (SafeTimer::timer_thread()+0x187) [0x7f4155e76397]
11: (SafeTimerThread::entry()+0xd) [0x7f4155e7798d]
12: (()+0x76db) [0x7f4154c9a6db]
13: (clone()+0x3f) [0x7f4153e7b88f]
You can get detail message from http://qa-proxy.ceph.com/teuthology/kchai-2018-08-27_08%3A20%3A01-rados-wip-kefu-testing-2018-08-27-1202-distro-basic-mira/2951064/.
@jecluis   This bug is like issue http://tracker.ceph.com/issues/24612. And Can you look at this and confirm #23742  whether also fix this bug?   Thanks!

**Comment by liewegas** [comment]
Let's rebase this (ceph_assert change) and repeat the test to see if it goes away..

**Comment by majianpeng** [comment]
@liewegas . rebased .

**Comment by majianpeng** [comment]
@tchaikov . Thanks very much!

**Comment by tchaikov** [comment]
not sure if this PR is relevant to the test failures, as quite a few PGs were stuck in "creating". i didn't look into the OSD log closely.

**Comment by majianpeng** [comment]
@tchaikov . I found the root reason. cause by 86c50c4 which omit a callback make pg can't activate and slow request occur.

**Comment by tchaikov** [comment]
@majianpeng needs rebase.

**Comment by majianpeng** [comment]
@tchaikov . update.

**Comment by tchaikov** [comment]
```xml
<stack>
    <frame>
      <ip>0xA96B12</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>BlueStore::queue_transactions(boost::intrusive_ptr&lt;ObjectStore::CollectionImpl&gt;&amp;, std::vector&lt;ObjectStore::Transaction, std::allocator&lt;ObjectStore::Transaction&gt; &gt;&amp;, boost::intrusive_ptr&lt;TrackedOp&gt;, ThreadPool::TPHandle*)</fn>
      <dir>/usr/src/debug/ceph-14.0.0-3041-gdadd632/src/os/bluestore</dir>
      <file>BlueStore.cc</file>
      <line>9831</line>
    </frame>
    <frame>
      <ip>0x6AB31F</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ObjectStore::queue_transaction(boost::intrusive_ptr&lt;ObjectStore::CollectionImpl&gt;&amp;, ObjectStore::Transaction&amp;&amp;, boost::intrusive_ptr&lt;TrackedOp&gt;, ThreadPool::TPHandle*)</fn>
      <dir>/usr/src/debug/ceph-14.0.0-3041-gdadd632/src/os</dir>
      <file>ObjectStore.h</file>
      <line>1457</line>
    </frame>
    <frame>
      <ip>0x636527</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSD::dispatch_context_transaction(PG::RecoveryCtx&amp;, PG*, ThreadPool::TPHandle*)</fn>
      <dir>/usr/src/debug/ceph-14.0.0-3041-gdadd632/src/osd</dir>
      <file>OSD.cc</file>
      <line>8719</line>
    </frame>
```

http://pulpito.ceph.com/kchai-2018-09-11_07:54:44-rados-wip-kefu2-testing-2018-09-10-1716-distro-basic-smithi/3005182/

and some other failures might also related.

**Comment by majianpeng** [comment]
@tchaikov . Thanks very much.

**Comment by majianpeng** [comment]
@tchaikov . Fix the valgrind issue.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-09-14_03:44:53-rados-wip-kefu-testing-2018-09-14-0938-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-15_11:45:38-rados-wip-kefu-testing-2018-09-14-0938-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-16_00:58:57-rados-wip-kefu-testing-2018-09-14-0938-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-16_11:53:24-rados-wip-kefu-testing-2018-09-14-0938-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-16_14:12:08-rados-wip-kefu-testing-2018-09-14-0938-distro-basic-mira/

the failed tests were caused by #19454 .

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -9857,8 +9857,8 @@ int BlueStore::queue_transactions(
     c->complete(0);
   }
   if (on_applied.size()) {
-    if (txc->ch->commit_queue)
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: please squash this change into 4ca3ccb21cb89007e54ae760c055ceee912a9a0f . in general, we don't fix a fix in the same PR.

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -9873,8 +9856,11 @@ int BlueStore::queue_transactions(
   for (auto c : on_applied_sync) {
     c->complete(0);
   }
-  for (auto c : on_applied) {
-    finishers[osr->shard]->queue(c);
+  if (on_applied.size()) {
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: nit, might want to put `if (!on_applied.empty())` if we don't care about the "size" here.

**Comment by tchaikov** [comment]
rerunning at http://pulpito.ceph.com/kchai-2018-09-18_08:09:08-rados-wip-22739-distro-basic-smithi/

**Comment by majianpeng** [comment]
@tchaikov . Still faied for the same reason:
2018-09-18T09:50:12.275 INFO:tasks.workunit.client.0.smithi154.stderr:mkfs failed: (22) Invalid argument
2018-09-18T09:50:12.277 INFO:tasks.workunit.client.0.smithi154.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test_kvstore_tool.sh:1: test_ceph_kvstorrm -fr /tmp/cephtool.q2a
2018-09-18T09:50:12.278 INFO:tasks.workunit:Stopping ['cephtool'] on client.0...
2018-09-18T09:50:12.278 INFO:teuthology.orchestra.run.smithi154:Running: 'rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0'
2018-09-18T09:50:12.432 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):

But for bluestore mkfs failed: (22) Invalid argument. THe code don't affect those code. @tchaikov , Could you give some suggestion? Thanks!

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-09-18_10:05:03-rados-master-distro-basic-smithi/ i am rerunning the same test on master again.

**Comment by liewegas** [comment]
Have we repeated the original performance test recently?  There have been a lot of changes to the code here and I would like to see that we are still seeing a sufficient benefit!

**Comment by majianpeng** [comment]
@tchaikov .  In rhel it failed But for Ubuntu i can passed. So i create a new #24167 which add --debug for ceph-objectstore-tool to get detail message.

**Comment by tchaikov** [comment]
@majianpeng before we devote more resources to this improvement, could you repeat the benchmark to address Sage's concern?

**Comment by majianpeng** [comment]
@tchaikov . I'm doing now by sage's  request.

**Comment by majianpeng** [comment]
| 4k-8osdthread-1finisher | 4k-8osdthread-8finisher | 4k-8osdthread-with-patch | 4k-16osdthread-1finisher | 4k-16osdthread-16finisher | 4k-16osdthread-with-patch
-- | -- | -- | -- | -- | -- | --
IOPS(K)(P3700 + First version+PG=128) | 51 | 51.7 | 49.7 | 50.6 | 54.9 | 58.6
IOPS(K)(P3700 + Current version+PG=128) | 62.1 | 60.2 | 59 | 59.3 | 59.2 | 63.3
IOPS(K)(Optane + current Version+PG=128) | 61.8 | 62.5 | 59.8 | 60.4 | 65.7 | 68.4
IOPS(K)(P3700 + Current version+PG=16) | 31.7 | 57.9 | 59 | 26.6 | 57.2 | 64.4
IOPS(K)((Optane + Current version+PG=16) | 37.3 | 62.7 | 60 | 37.2 | 66.8 | 68.4

- Decrease in growth for 4k-16osdthread-with-patch/4k-16osdthread-1finisher:  from firs version(58.6/50.6=1.15) to (current version) (63.3/59.3=1.07). But for Optane(faster device than P3700), there is still 13% performance improved. (68.4/60.4)

- for PG=16, we see about double performance dropped for one finisher. But at  this condition, this PR can improve performance(about 100%) than one-finisher and multi-finisher.

@liewegas  From the results, i think this PR  still deserves to be merged.

**Comment by liewegas** [comment]
Sounds good!  Let's proceed

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-09-20_04:43:32-rados-wip-kefu-testing-2018-09-19-1821-distro-basic-smithi/

File: src/common/Finisher.h [hunk 7]
```diff
@@ -184,4 +184,46 @@ class C_OnFinisher : public Context {
...
+      cond.Signal();
+      mutex.Unlock();
+    }
+
+    ls.clear();
+  }
+
+  void swap(list<Context *>& ls) {
+    ls.clear();
+    std::lock_guard<std::mutex> l(q_mutex);
```
Comment (tchaikov) [inline src/common/Finisher.h]: use `scoped_lock` if you please.

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -8934,7 +8928,10 @@ void BlueStore::_txc_committed_kv(TransContext *txc)
   {
     std::lock_guard<std::mutex> l(txc->osr->qlock);
     txc->state = TransContext::STATE_KV_DONE;
-    finishers[txc->osr->shard]->queue(txc->oncommits);
+    if (txc->ch->commit_queue)
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: https://github.com/ceph/ceph/pull/22739#discussion_r217981442

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -9873,8 +9856,11 @@ int BlueStore::queue_transactions(
   for (auto c : on_applied_sync) {
     c->complete(0);
   }
-  for (auto c : on_applied) {
-    finishers[osr->shard]->queue(c);
+  if (!on_applied.empty()) {
+    if (c->commit_queue)
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: https://github.com/ceph/ceph/pull/22739#discussion_r217981442

File: src/os/bluestore/BlueStore.h [hunk 10]
```diff
@@ -2410,6 +2406,8 @@ class BlueStore : public ObjectStore,
 
   CollectionHandle open_collection(const coll_t &c) override;
   CollectionHandle create_new_collection(const coll_t& cid) override;
+  void set_collection_commit_queue(const coll_t& cid,
+				   class ContextQueue *commit_queue) override;
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981026

File: src/os/bluestore/BlueStore.h [hunk 11]
```diff
@@ -1368,6 +1368,7 @@ class BlueStore : public ObjectStore,
 
     //pool options
     pool_opts_t pool_opts;
+    class ContextQueue *commit_queue;
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981026

File: src/os/filestore/FileStore.h [hunk 12]
```diff
@@ -531,6 +531,9 @@ class FileStore : public JournalingObjectStore,
 
   CollectionHandle open_collection(const coll_t& c) override;
   CollectionHandle create_new_collection(const coll_t& c) override;
+  void set_collection_commit_queue(const coll_t& cid,
+      class ContextQueue *commit_queue) override {
```
Comment (tchaikov) [inline src/os/filestore/FileStore.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981026

File: src/os/kstore/KStore.h [hunk 13]
```diff
@@ -443,6 +443,9 @@ class KStore : public ObjectStore {
 
   CollectionHandle open_collection(const coll_t& c) override;
   CollectionHandle create_new_collection(const coll_t& c) override;
+  void set_collection_commit_queue(const coll_t& cid,
+      class ContextQueue *commit_queue) override {
```
Comment (tchaikov) [inline src/os/kstore/KStore.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981026

File: src/os/memstore/MemStore.h [hunk 14]
```diff
@@ -321,6 +321,11 @@ class MemStore : public ObjectStore {
     return get_collection(c);
   }
   CollectionHandle create_new_collection(const coll_t& c) override;
+
+  void set_collection_commit_queue(const coll_t& cid,
+      class ContextQueue *commit_queue) override {
```
Comment (tchaikov) [inline src/os/memstore/MemStore.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981026

File: src/osd/OSD.h [hunk 15]
```diff
@@ -1774,7 +1776,16 @@ class OSD : public Dispatcher,
       auto &&sdata = osd->shards[shard_index];
       ceph_assert(sdata);
       Mutex::Locker l(sdata->shard_lock);
-      return sdata->pqueue->empty();
+      if (thread_index < osd->num_shards)
```
Comment (tchaikov) [inline src/osd/OSD.h]: https://github.com/ceph/ceph/pull/22739#discussion_r217981442

**Comment by tchaikov** [comment]
> How to handle https://github.com/ceph/ceph/pull/24167. Close or?

yeah, please close it.

**Comment by majianpeng** [comment]
@tchaikov . update by your comments. Thanks very much!

**Comment by liewegas** [comment]
Wahoo!

**Comment by majianpeng** [comment]
@liewegas @tchaikov . Thanks very much for help make this be merged at finale.

**Comment by tchaikov** [comment]
tracked by http://tracker.ceph.com/issues/36099

**Comment by majianpeng** [comment]
@tchaikov . i'll check this now.

**Comment by liewegas** [comment]
@rzarzynski started looking at this yesterday too

**Pull request merged**